tests/Doctrine/Tests/DBAL/Functional/Schema/MySqlSchemaManagerTest.php

```diff
@@ -325,4 +332,244 @@ public function testListFloatTypeColumns()
+    {
+        $this->_conn->query('DROP TABLE IF EXISTS test_column_defaults_with_create');
+        $sql = "
+            CREATE TABLE test_column_defaults_with_create (
+                col1 VARCHAR(255) NULL DEFAULT 'O''Connor\'\"',              
+                col2 VARCHAR(255) NOT NULL DEFAULT 'O''Connor\'\"'                                
+                );
+        ";
+        $this->_conn->query($sql);
+
+        $onlineTable = $this->_sm->listTableDetails("test_column_defaults_with_create");
+        self::assertSame("O'Connor'\"", $onlineTable->getColumn('col1')->getDefault());
+        self::assertSame("O'Connor'\"", $onlineTable->getColumn('col2')->getDefault());
+
+        $table = new Table("test_column_defaults_no_diff");
+        $table->addColumn('col1', 'string', ['notnull' => false, 'default' => "O'Connor'\""]);
+        $table->addColumn('col2', 'string', ['notnull' => true, 'default' => "O'Connor'\""]);
+
+        $comparator = new Comparator();
+        $diff = $comparator->diffTable($table, $onlineTable);
+        self::assertFalse($diff);
+    }
+
+    public function testColumnDefaultCurrentTimestamp() : void
+    {
+        $platform = $this->_sm->getDatabasePlatform();
+
+        $table = new Table("test_column_defaults_current_timestamp");
+
+        $currentTimeStampSql = $platform->getCurrentTimestampSQL();
+
+        $table->addColumn('col_datetime', 'datetime', ['notnull' => true, 'default' => $currentTimeStampSql]);
+        $table->addColumn('col_datetime_nullable', 'datetime', ['default' => $currentTimeStampSql]);
+
+        $this->_sm->dropAndCreateTable($table);
+
+        $onlineTable = $this->_sm->listTableDetails("test_column_defaults_current_timestamp");
+        self::assertSame($currentTimeStampSql, $onlineTable->getColumn('col_datetime')->getDefault());
+        self::assertSame($currentTimeStampSql, $onlineTable->getColumn('col_datetime_nullable')->getDefault());
+
+        $comparator = new Comparator();
+
+        $diff = $comparator->diffTable($table, $onlineTable);
+        self::assertFalse($diff, "Tables should be identical with column defaults.");
+    }
+
+    public function testColumnDefaultsAreValid() {
+
+        $table = new Table("test_column_defaults_are_valid");
+
+        $currentTimeStampSql = $this->_sm->getDatabasePlatform()->getCurrentTimestampSQL();
+        $table->addColumn('col_datetime', 'datetime', ['default' => $currentTimeStampSql]);
+        $table->addColumn('col_datetime_null', 'datetime', ['notnull' => false, 'default' => null]);
+        $table->addColumn('col_int', 'integer', ['default' => 1]);
+        $table->addColumn('col_neg_int', 'integer', ['default' => -1]);
+        $table->addColumn('col_string', 'string', ['default' => 'A']);
+        $table->addColumn('col_decimal', 'decimal', ['scale' => 3, 'precision' => 6, 'default' => -2.3]);
+        $table->addColumn('col_date', 'date', ['default' => '2012-12-12']);
+
```

Comment:
It'd be nice to have a string containing characters which require escaping. It would cover the changes in `AbstractPlatform::getDefaultValueDeclarationSQL()` in a functional way and could be executed across supported platforms.

Reply:
Yes, but I'm not sure yet about keeping default. I'll tell you later. SQLServer quoteStringLiterals is buggy. Need to talk about this.

Reply:
@morozov I've dropped quoting feature in favour of working on #2850

Reply:
There's still code in `getMariaDb1027ColumnDefault()` which parses an escaped string. Not sure how correct the parsing logic is, given that the default value may contain escaped characters.

Reply:
@morozov, I assume you're refering to the `stripslashes` in `getMariaDb1027ColumnDefault()`.

```
stripslashes(preg_replace('/^\'(.*)\'$/', '$1', $columnDefault));`  
```
I can remove it. Let me know, so I can be sure that's what you're pointing out ?

Thanks

Reply:
@morozov , FYI I was pretty busy lately... But once I've finished my current work, I really plan to either work on #2850 or make a new P/R to handle escaping in defaults.

Reply:
Yes, I was referring to that. As far as I understand, it cannot be removed because it's a requirement for the MariaDB platform support.

It looks like this ticket very much depends on #2850 which implements un-quoting for other platforms. Would it make sense to resolve that one first and then implement MariaDB with the logic of handling default expressions in place?

I'm not sure if quoting default values in DDL and un-quoting default expressions are parts of the same feature, and if not, why does #2850 have to implement un-quoting?

Reply:
> As far as I understand, it cannot be removed because it's a requirement for the MariaDB platform support.

Fortunately not a requirement ;) I've removed it for clarity... it will have to be added in #2850 

> Would it make sense to resolve that one first and then implement MariaDB with the logic of handling default expressions in place?

The two P/R can be shipped separately. MariaDB 10.2 support should be shipped as soon as possible. 

> I'm not sure if quoting default values in DDL and un-quoting default expressions are parts of the same feature, and if not, why does #2850 have to implement un-quoting?

I'll give you more detail on #2850... But my feelings on this would be:

1. Ship MariaDB 10.2 support
2. Ship @powerkiki quoting default **values**, #2850 (it's already quite a work to figure out platform support)
3. Create a new P/R allowing to distinguish values and expressions (maybe introduce a new type).

Let me know if we see it the same way

Reply:
@belgattitude I agree with you on your step-by-step approach. We should not try do do everything at once, or else we might never finish anything.

Reply:
> Fortunately not a requirement ;) I've removed it for clarity... it will have to be added in #2850

Isn't it what's still implemented in `MySqlSchemaManager::getMariaDb1027ColumnDefault()`? You're quoting it in the ticket summary:

> In order to allow [expressions](https://jira.mariadb.org/browse/MDEV-10134) as default values and distinguish them from literals, Mariadb now quotes default values in `information_schema.column` table.

Do you mean handling some other expressions is not a requirement? Because a quoted literal, non-quoted `CURRENT_TIMESTAMP()` and `NULL` (string) they all _are expressions_.

Given that this is one of the major differences between currently supported MySQL and MariaDB 10.2.7, shouldn't this logic be extracted into `MariaDb1027SchemaManager`?

Reply:
The `stripslashes()` shouldn't have been removed from the code because this is (presumably) how quoted quotes should be handled in the default values. Alongside with the literals not containing special characters (like `'A'`), this same test should cover literals _with_ special characters. This will show the need to parse the quoted literal expression.

Reply:
@morozov, got lost somewhere... :) 

> Isn't it what's still implemented in MySqlSchemaManager::getMariaDb1027ColumnDefault()? You're quoting it in the ticket summary:

Probably my english is wrong somewhere... I mean default literal are quoted in MariaDB102. Support for escape characters in default literals (non-existent yet) will be covered in #2850, not in this P/R. I'll help @powerkiki to be sure it works for MariaDB 10.2 as well. 

> Do you mean handling some other expressions is not a requirement? Because a quoted literal, non-quoted CURRENT_TIMESTAMP() and NULL (string) they all are expressions.

Have a look on how it's done in the current dbal implementation :
https://github.com/doctrine/dbal/blob/09e072dee1457bcdf3cae4be60b18854f8829992/lib/Doctrine/DBAL/Platforms/AbstractPlatform.php#L2288

Handling expressions is not required in the context of this P/R (mariadb 10.2 support). This P/R just map everything back to what is expected by the current dbal implementation.   

To the question, can we extend how doctrine handle expressions (i.e to support *arbitrary* expressions like `concat()`...): Yes it's nice to think about it... But in a totally new P/R, not this one.

> Given that this is one of the major differences between currently supported MySQL and MariaDB 10.2.7, shouldn't this logic be extracted into MariaDb1027SchemaManager

No this P/R maps back to what the current implementation expects. But when support for *arbitrary* expressions are implemented, yes it must be done. *(arbitray is not a good name, but I don't have better for now... Just to say expressions, except TIMESTAMP(), NULL... )*

> The stripslashes() shouldn't have been removed from the code because this is (presumably) how quoted special characters should be handled in the default values. 

This will be handled in #2850 ... That's why I removed it (see my questions in the P/R discussion), I can re-add the `stripslashes`.. but will be a never ending story when you'll ask to provide tests ;) 

> Alongside with the literals not containing special characters (like 'A'),** this same test should cover literals with special characters**. 

I just removed it, in favor of #2850, where I'm able to focus and contribute. Let's keep it that way.  

> This will show the need to parse the quoted literal expression.

Euh, literal expressions are not quoted... That's exactly how MariaDB 10.2, Postgres... are doing to distinguish them from literal values. 

Support for literal expressions (what I called *arbitrary* expression) in a future P/R.

Sorry if my english is not clear, I tried my best ;) But thanks for persisting on this. 

Is that more clear to you ?

Reply:
@belgattitude there's nothing wrong with your English. However it, looks like we're not on the same page regarding the terminology. Let's establish some common ground.

What I mean by expressions, is any string which represents a valid part of SQL. Strings `1`, `'2'`, `NULL`, `'foo''bar'`, `'foo\'bar'` and `CURRENT_TIMESTAMP()` are all valid SQL expressions for MariaDB. And any string returned in the `COLUMN_DEFAULT` column (except the `NULL` value), is a valid SQL expression. Correct me if I'm wrong.

Everything above besides the `CURRENT_TIMESTAMP()` are [literal](https://docs.oracle.com/cd/B19306_01/server.102/b14200/sql_elements003.htm) (string, numeric, dates, etc) expressions. They represent [literals](https://en.wikipedia.org/wiki/Literal_(computer_programming)) (fixed values).

We indeed don't need to be able to handle arbitrary expressions like `CONCAT('foo', 'bar')`  to support MariaDB but we do need to be able to handle literal expressions. Because this how MariaDB represents default _values_.

> Euh, literal expressions are not quoted...

They are:
```
Welcome to the MySQL monitor.  Commands end with ; or \g.
Server version: 5.5.5-10.2.10-MariaDB-10.2.10+maria~jessie mariadb.org binary distribution

mysql> CREATE TABLE test (
    ->   value VARCHAR(32) DEFAULT 'foo\'bar'
    -> );
Query OK, 0 rows affected (0.03 sec)

mysql> SELECT COLUMN_NAME, COLUMN_DEFAULT FROM information_schema.COLUMNS WHERE TABLE_NAME = 'test';
+-------------+----------------+
| COLUMN_NAME | COLUMN_DEFAULT |
+-------------+----------------+
| value       | 'foo''bar'     |
+-------------+----------------+
2 rows in set (0.00 sec)

mysql> INSERT INTO test VALUES ();
Query OK, 1 row affected (0.01 sec)

mysql> 
mysql> SELECT * FROM test;
+---------+
| value   |
+---------+
| foo'bar |
+---------+
1 row in set (0.00 sec)
```
As you can see, the default value of `(string) foo'bar` is represented as a SQL literal expression: `'foo''bar'`. The string value is quoted (surrounded by quotes) and escaped (special characters are encoded in the way they lose their special meaning).

In order to support schema introspection for MariaDB, we should be able to understand that `'foo''bar'` retrieved from the introspection query is a _literal expression_ which represents the `(string) foo'bar` _value_. 

Right now, I don't see any code in `getMariaDb1027ColumnDefault()` or tests which would cover those inner quote characters.

> Support for literal expressions (what I called arbitrary expression) in a future P/R.

Looks like the "literal" word is misused. It's not an arbitrary expression.

> Support for escape characters in default literals (non-existent yet) will be covered in #2850.

Same here. Literal expressions representing defaults is the major difference between MariaDB and MySQL. If you want to omit/postpone handling escaping in literal expressions, what exactly does "MariaDB support" mean in your understanding?

Reply:
@morozov , thanks

> Terminology...

Thank you to point out. So if i get it right.... there's no such thing like 'literal expressions'... It's either a literal (`NULL`, `A`, `2012-12-12`) or an expression (`CURRENT_TIMESTAMP()`, `CURRENT_TIME()`, `CONCAT(x, y)`, imagine also `nextval`...). 

> **Literal expressions** representing defaults is the major difference between MariaDB and MySQL

Sorry that's where I'm lost again :)  Is that `default literal values` vs `default literal expressions` or simply `default values` vs `default expressions`.

> mysql> CREATE TABLE test (value VARCHAR(32) DEFAULT 'foo\'bar'...

Yes... This P/R fixes exactly: `CREATE TABLE test (value VARCHAR(32) DEFAULT 'foobar'` (this very simple case breaks the schema manager with MariaDB10.2)... not `CREATE TABLE test (value VARCHAR(32) DEFAULT 'foo\'bar'`. If you want to support escaping, the proposal is #2850 (all platforms).

> I don't see any code in getMariaDb1027ColumnDefault() or tests which would cover those inner quote characters.

Yes I implemented it and removed it in favor of of #2850. Current doctrine implementation does not really support escaping in defaults values (and/or literals). The fix is super easy... but should be done for all platforms... (I only refer to escaping here). See the thread in #2850, It covers much more cases (complete escape sequences...).

> what exactly does "MariaDB support" mean in your understanding?

I tried to keep an updated explanation in the P/R description. Can you have a look to the first 6 points ? and why I removed two of them (all implemented before, see the discussion threads where I asked if I could do it that way).

Could you just tell me if you agree we can do it in 3 independent steps (very hard to me to have an answer to that):

1. Ship this P/R 'as it is'. Scope: see the 6 steps in the P/R description. The one important is the number 6. This is where we support the MariaDB changes (+/- a default value appears in information_schema like `'Hello'` instead of `Hello` in Mysql, `current_timestamp()` (UNQUOTED) is stored instead of MySQL `CURRENT_TIMESTAMP`, `'NULL'` is stored instead of `null`...). This part does not impact other platforms. (new platform support). It will bring support for MariaDB 10.2 with +/- the same features as current MySQL (without escaping) and fixes 

2. Handle escaping of defaults in #2850. Could be shipped before or after... it's independent. It should be done for all platforms. In the MariaDB 102 context, it will add support for setting default values .

3.  Create a new P/R to support expressions.

Sorry to re-ask all the time the same question. But it is basically this answer I need and could not have (see dicussion thread). Once I get it  all your questions will find answers.

Reply:
@belgattitude literals _are_ expressions. So yes, "literal expression" is a kind of tautology.

> Sorry that's where I'm lost again :)  Is that `default literal values` vs `default literal expressions` or simply `default values` vs `default expressions`.

In general, it's just a default (noun), a default (adj.) thingy. Depending on the platform, it's either the value itself (MySQL) or the expression which defines how the default value is calculated (MariaDB). For MariaDB it can be a literal (`'foo''bar'`) which always evaluates to the same value or an arbitrary expression (`CURRENT_TIMESTAMP()`).

> Yes I implemented it and removed it in favor of #2850.

I think now I get it. Your position is that as long as we cannot declare the default value with special characters, we don't have to be able to introspect them correctly as well. And #2850 is trying address both the declaration and introspection aspects.

My point of view on this is different, and that seems to be the source of our mutual misunderstanding. I think that these aspects are related but independent. The DBAL can be used only for declaration or only for introspection. It's not a black box.

There may be some existing tests which cover both aspects at once, that's why the author decided to attack both of them in a single PR. If so, I do agree it's a valid approach since it keeps the build green.

> Could you just tell me if you agree we can do it in 3 independent steps (very hard to me to have an answer to that)

I agree we can break all the work down into these 3 steps but I'm not sure (not disagreeing) they are independent. As for the shipping, I don't have an opinion because I'm not a MariaDB user.

Right now, the MySQL platform supports introspection of default values with special characters but MariaDB doesn't. From this standpoint, currently, there's no parity between supported features for MySQL and MariaDB.

So, the bottom line is, I don't mind leaving handling defaults as is for now if other maintainers are okay with that. Thanks for bearing with me and for your hard work.

Reply:
> So, the bottom line is, I don't mind leaving handling defaults as is for now if other maintainers are okay with that. Thanks for bearing with me and for your hard work.

Great. Thanks a lot

Reply:
> I think now I get it. Your position is that as long as we cannot declare the default value with special characters, we don't have to be able to introspect them correctly as well. And #2850 is trying address both the declaration and introspection aspects.

Yes and no. I'm personnally in favour of keeping the escaping (I can easily re-eable them). But I'm not able to provide functional tests (in a good way) without impacting other plaforms too. (the `quoteStringLiterals` should be done in AbstractPlatform and I try to avoid regressions or bc in this P/R as much as possible). 

I can of course rea-dd the escaping. Tell me if you prefer this way (I prefer it too) ? but I cannot provide functionnal tests (support for backslashes in default is still inexistent, try to create a default value containing a backslash - it will trigger continuous schema diffs).

Reply:
I'm in favor of having (un?)escaping implemented here as well. You can create the schema using SQL and then introspect it using the Schema manager. Once the other issue is fixed, this test can be reworked or removed in favor of the one which uses Schema manager in both cases and doesn't depend on the platform.

Reply:
That's exactly the kind of answer I need, thanks :) Have a look to my latest commit.